planner/core/rule_join_reorder_greedy.go

```diff
@@ -210,13 +210,24 @@ func (s *joinReOrderGreedySolver) optimizeRecursive(ctx sessionctx.Context, p Lo
 				return nil, err
 			}
 		}
-		groupSolver := &joinReorderGreedySingleGroupSolver{
-			ctx:          ctx,
-			curJoinGroup: curJoinGroup,
-			eqEdges:      eqEdges,
-			otherConds:   otherConds,
+		if len(curJoinGroup) > ctx.GetSessionVars().TiDBOptJoinOrderAlgoThreshold {
+			greedySolver := &joinReorderGreedySingleGroupSolver{
+				ctx:          ctx,
+				curJoinGroup: curJoinGroup,
+				eqEdges:      eqEdges,
+				otherConds:   otherConds,
+			}
+			p, err = greedySolver.solve()
+			if err != nil {
+				return nil, err
+			}
+			return p, nil
+		}
+		dpSolver := &joinReorderDPSolver{
+			ctx: ctx,
 		}
-		p, err = groupSolver.solve()
+		dpSolver.newJoin = dpSolver.newJoinWithConds
```

Comment:
`newJoin` can be other function pointers later? why not removing `newJoin` field and directly using the function?